lib/puppet/type/agent_upgrade.rb

```diff
@@ -0,0 +1,30 @@
+Puppet::Type.newtype(:agent_upgrade) do
+  ensurable
+
+  newparam(:name, :namevar => true)
+
+  newparam(:source) do
+    newvalues /^((https?|puppet):\/\/)|\\\\\S+\\\w+|[a-z]:\\/i
+    validate do |value|
+      if value !~ /^((https?|puppet):\/\/)|\\\\\S+\\\w+|[a-z]:\\/i
+        fail "Please provide a valid http(s), puppet or unc path"
+      end
+      if value =~ /^https?/ && value !~ URI::regexp
+        fail("Please provide a valid URI")
```

Comment:
Probably better to call `uri = URI.parse(value)` and them verify `uri.scheme` is one of the supported values. That said, I don't think we can support `puppet` URIs, as the agent does some magic converting `puppet:///modules/mymodule/file` into an HTTP request.